### ISSUE TITLE
RT #114541: Strange dropping of 0's with _ versions

### ISSUE THREAD

**Issue by briandfoy** [briandfoy/perl-version#10]
From KENTNL in [#114541 on rt.cpan.org](https://rt.cpan.org/Ticket/Display.html?id=114541)

Probably shouldn't be setting version to have an underscore with this ... but ... either way, this was surprising.

    perl-reversion -set 1.000005_001

...

    +our $VERSION   = '1.000005_01';

uuuuh.

**Comment by briandfoy** [comment]
This is indeed surprising, and if you ask for something, that's what you should get.

perl-revision uses `_%02d` for formatting alpha versions, which in `Perl::Version::set` happens separately from the rest of the version string. `Perl::Version` uses `_%03d` by default.

One of the problems is that the output of perl-reversion stringifies the object, and that code follows a different path than `set`. 

This line in `perl-reversion` outputs a different version than the one it's going to set:

      note( "Setting version to $new_version\n" );  # 1.000005_001

Changing it to this outputs the version it will set:

      note( "Setting version to " . $new_version->numify . "\n" ); # 1.000005_01

There's another problem. The first version that Perl::Version gets sets the format. If you are starting with a user release, like `1.015`, there is no alpha component so you get the default format of `_%02d`. `set` takes an existing object and replaces parts of it but does not update the format length for the new alpha part.

This is also a problem for `bump`, but in a different way.

**Comment by briandfoy** [comment]
I think whatever I did to fix this also takes care of dump. I added some tests to *t/40.perl-revision.t*, and I think I covered the cases that matter.

One of the issues I don't want to clear up is how many parts of the code aren't using the same paths to go from one place to the same place. These bugs show up in different code paths because there isn't a coherent code path. I also don't like the idea of one object surgical replacing parts of another object. Immutable objects would be much better here. But, I'm not going to do that work.
